File: hasher.js
var crypto = require('crypto');
var _ = require('lodash');
var util = require('util');

var hash = function(hash_type,data,options){
	if (!options) {		
		options = {}
	}
	_.defaults(options,{
		"sort_keys": true,
		"sort_arrays": false,
	});	
	var h = crypto.createHash(hash_type);

	s = "";
	if(_.isArray(data)){
		if (options.sort_arrays) {
			data.sort();
		}
		data.forEach(function(i){
			s += hash(hash_type,i,options);
		})
	} else if (_.isString(data)){
		s = data;
	} else if (_.isNumber(data)) {
		s = util.format("%d",data);
	} else if (_.isPlainObject(data)) {
		var ks = _.keys(data);
		if(options.sort_keys) {
			ks.sort();
		}
		ks.forEach(function(k){
			s += k + hash(hash_type,data[k],options);
		})
	} else {
		console.log(typeof data);
	}

	return h.update(s).digest('hex');;
}

module.exports.hash = hash;

File: hasher.py
import hashlib

def hash(hash_type,data,sort_arrays=False,sort_keys=True):
	h = hashlib.new(hash_type)

	s = ""
	if isinstance(data,list):
		sa = []
		for i in data:
			sa.append(hash(hash_type,i,sort_arrays=False,sort_keys=True))
		if sort_arrays:
			sa.sort()
		s = "".join(sa)

	elif isinstance(data,str) or isinstance(data,unicode):
		s = data;
	elif isinstance(data,int) or isinstance(data,float):
		s = str(data);
	elif isinstance(data,dict):
		ks = data.keys()
		if sort_keys:
			ks.sort()

		for k in ks:
			s += k + hash(hash_type,data[k],sort_arrays=False,sort_keys=True)

	else:
		print type(data);

	# print s
	h.update(s)
	return h.hexdigest()
